nightguard/MainViewController.swift

```diff
@@ -181,40 +174,54 @@ class MainViewController: UIViewController {
         return UIStatusBarStyle.lightContent
     }
     
-    func startTimer() {
+    
+    // MARK: Notification
+    @objc func applicationDidEnterBackground(_ notification: Notification) {
+        timer.invalidate()
+        AlarmSound.stop()
+    }
+    
+    @objc func applicationWillEnterForeground(_ notification: Notification) {
+        
+        // If there is already a snooze active => we don't have to fear that an alarm
+        // would be played.
+        if !AlarmRule.isSnoozed() {
+            // snooze the alarm for 15 Seconds in order to retrieve new data
+            // before playing alarm
+            // Otherwise it could be the case that the app immediately plays
+            // an alarm sound without giving the app the chance to reload
+            // current data
+            AlarmRule.snoozeSeconds(15)
+            self.updateSnoozeButtonText()
+        }
+
+        startTimer()
+        doPeriodicUpdate(forceRepaint: true)
+    }
+
+    fileprivate func startTimer() {
         timer = Timer.scheduledTimer(timeInterval: timeInterval,
                                      target: self,
                                      selector: #selector(MainViewController.timerDidEnd(_:)),
                                      userInfo: nil,
                                      repeats: true)
     }
     
-    @objc func safetyResetTimerDidEnd(_ timer: Timer) {
-        timer.invalidate()
-        startTimer()
-    }
-    
     // check whether new Values should be retrieved
     @objc func timerDidEnd(_ timer:Timer) {
         
-        if AlarmRule.isAlarmActivated(NightscoutCacheService.singleton.getCurrentNightscoutData(), bloodValues: NightscoutCacheService.singleton.getTodaysBgData()) {
-            
-            // Play the sound only if foreground => otherwise this won't work at all
-            // and the sound will only play right when opening the application :-/
-            let state = UIApplication.shared.applicationState
-            if state == UIApplicationState.active {
-                AlarmSound.play()
-            }
-        } else {
-            AlarmSound.stop()
+        DispatchQueue.main.async { [unowned self] in
```

Comment:
Why aren't you using your Util-Method here?

Reply:
Do you ask about dispatchOnMain? I think this code was written earlier than the dispatchOnMain idea was implemented. Now, as the code changed for a few times in this area, I think the async call is not really needed anymore, the load methods from doPeriodicUpdate are already async calls. And the timer call comes always on the main thread.